aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -12,47 +12,90 @@
- *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      buffptr member corresponding to char_offset. This value is only set when a matching char_offset is found
  *      in aesd_buffer.
  * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
  * NULL if this position is not available in the buffer (not enough data is written).
  */
 struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
-            size_t char_offset, size_t *entry_offset_byte_rtn )
+            size_t char_offset, size_t *entry_offset_byte_rtn)
 {
-    /**
-    * TODO: implement per description
-    */
+    // Check if inputs are valid pointers
+    if (buffer == NULL || entry_offset_byte_rtn == NULL) {
+        return NULL;
+    }
+
+    // Initialize variables
+    int lengthOfString = 0;
+    int total_entries_visited = 0;
+    char_offset++; // Adjusting for zero-referenced index
+
+
+    // First check the out_offs buffer entry and check the length of buffer. 
+    // 1) If length is < actual offset move to next buffer and reduce the length -= strlen(buffer); Go to next circular before pos
+    // else 2) retrun len -1 ;(which will be the pos of corresponding buffer)
+    // Iterate over the circular buffer
+    for (int seek_off = buffer->out_offs; total_entries_visited < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED; 
+         total_entries_visited++, seek_off = (seek_off + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) 
+    {
+
+        lengthOfString = strlen(buffer->entry[seek_off].buffptr);
+        if (lengthOfString < char_offset) {
+            char_offset -= lengthOfString;
+        } else {
+            *entry_offset_byte_rtn = char_offset - 1;
+            return &buffer->entry[seek_off];
+        }
+    }
+
+    *entry_offset_byte_rtn = (size_t)-1; // Indicate no valid entry found
     return NULL;
 }
 
 /**
 * Adds entry @param add_entry to @param buffer in the location specified in buffer->in_offs.
 * If the buffer was already full, overwrites the oldest entry and advances buffer->out_offs to the
-* new start location.
-* Any necessary locking must be handled by the caller
+* new start location. Any necessary locking must be handled by the caller.
 * Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
 */
 void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
 {
-    /**
-    * TODO: implement per description
-    */
+    // Check if inputs are valid pointers
+    if (buffer == NULL || add_entry == NULL) {
+        return;
+    }
+
+    // Add the entry to the buffer
+    buffer->entry[buffer->in_offs] = *add_entry;
+
+    // Check if buffer is full
+    if (buffer->in_offs == buffer->out_offs) {
```

Comment:
This will set full on the first call when the buffer only has an initial element, since both in and out offset are initialized to zero.  You should set this on a check after the increment instead.